services/src/main/java/org/keycloak/services/resources/ThemeResource.java

```diff
@@ -138,8 +141,17 @@ public Response getLocalizationTexts(@PathParam("realm") String realmName, @Quer
                     new KeySource((String) e.getKey(), (String) e.getValue())).collect(toList());
         }
 
+        if (!hasWords.isEmpty()) {
+            result = result.stream().filter(keySource -> hasWords.stream()
+                    .anyMatch(w -> keySource.getValue().toUpperCase(locale).contains(w.toUpperCase(locale)))).collect(toList());
+        }
+
+        if (max != 0) {
+            result = result.stream().skip(first).limit(max).collect(toList());
```

Comment:
In Keycloak REST API it is common to have `first` and `max` settable independently.

This part of code converts a stream to list and back several times which increases pressure on GC due to number of created fresh objects. Instead, keep the stream only and collect at the end. That way it is also possible to have `.skip` and `.limit` applied selectively when the `first` / `max` are non-null, and similarly `hasWords` parameter.